relay-server/src/actors/processor.rs

```diff
@@ -75,9 +78,191 @@ mod span;
 
+/// Creates the block only if used with `processing` feature.
+///
+/// Provided code block will be executed only if the provided config has `processing_enabled` set.
+macro_rules! if_processing {
+    ($config:expr, $if_true:block) => {
+        #[cfg(feature = "processing")] {
+            if $config.processing_enabled() $if_true
+        }
+    };
+}
+
 /// The minimum clock drift for correction to apply.
 const MINIMUM_CLOCK_DRIFT: Duration = Duration::from_secs(55 * 60);
 
+/// Describes the groups of the processable items.
+#[derive(Clone, Copy, Debug)]
+pub enum ProcessingGroup {
+    /// All the transaction related items.
+    ///
+    /// Includes transactions, related attachments, profiles.
+    Transaction,
+    /// All the items which require (have or create) events.
+    ///
+    /// This includes: errors, NEL, security reports, user and clients reports, some of the
+    /// attachments.
+    Error,
+    /// Session events.
+    Session,
+    /// Attachments which can be sent alone without any event attached to it in the current
+    /// envelope.
+    StandaloneAttachment,
+    UserReport,
+    Replay,
+    /// Crons.
+    CheckIn,
+    Span,
+    /// Fallback.
+    ///
+    /// Kept for forward compatibility.
+    Unknown,
+}
+
+impl ProcessingGroup {
+    /// Returns the group associated with the provided [`Envelope`].
+    ///
+    /// Based on the items, which the provided envelope contains, make the decision, which
+    /// group should be assigned.
+    ///
+    /// The decision made in the following order:
+    /// - if any of the items is a transaction - return group `Transaction`
+    /// - if the envelope contains only attachment items, and those attachments do not create
+    /// the event - we are dealing with `StandaloneAttachment`
+    /// - if there are ReplayEvent or ReplayRecording items - `Replay` group is assigned
+    /// - if any of the items require event - this is the `Error` group
+    /// - the rest of the items have dedicates groups: Session, UserReport, CheckIn,
+    /// Span
+    /// - the `Unknown` group is for forward-compatibility and assigned to the items which do
+    /// not fit into any of the groups mentioned above.
+    pub fn from_envelope(envelope: &Envelope) -> Self {
+        if envelope
+            .items()
+            .any(|item| item.ty() == &ItemType::Transaction)
+        {
+            return Self::Transaction;
+        }
+
+        if envelope
+            .items()
+            .any(|item| item.ty() == &ItemType::Attachment || item.ty() == &ItemType::FormData)
+            && !envelope.items().any(Item::creates_event)
+        {
+            return Self::StandaloneAttachment;
+        }
+
+        if envelope.items().any(|item| {
+            item.ty() == &ItemType::ReplayEvent || item.ty() == &ItemType::ReplayRecording
+        }) {
+            return Self::Replay;
+        }
+
+        if envelope.items().any(Item::requires_event) {
+            return Self::Error;
+        }
```

Comment:
I noticed one edge case here:

When a profile comes in without a transaction, `from_envelope` will return `Self::error`. But `process_error` does not call `profile::filter`, which takes care of dropping the profile!

https://github.com/getsentry/relay/blob/635a7cf051741f214a0c7cffd05f9593f1b65924/relay-server/src/actors/processor/profile.rs#L26-L27

I.e. we should return `ProcessingGroup::Transaction` for `ItemType::Profile`. There might be other cases like this, so maybe we should explicitly enumerate the item types that are allowed in each group, and send other items to the `Unknown` group. Together with the "safety net" I suggested below, this might help us catch bugs (if there are any).